File: js/games/MovementGameBase.js
import { GameBase } from '../engine/GameBase.js';

const WIN_LINES = [
  [0, 1, 2], [3, 4, 5], [6, 7, 8],
  [0, 3, 6], [1, 4, 7], [2, 5, 8],
  [0, 4, 8], [2, 4, 6],
];

// P1 at b3(1), a1(6), c1(8); P2 at a3(0), c3(2), b1(7); empty: a2(3), b2(4), c2(5)
const INITIAL_BOARD = [2, 1, 2, 0, 0, 0, 1, 2, 1];

const MAX_MOVES = 120;

export class MovementGameBase extends GameBase {
  get interactionMode() { return 'movement'; }

  get boardConfig() { return { type: 'tapatan', size: 3 }; }

  getInitialState() {
    return {
      board: [...INITIAL_BOARD],
      currentPlayer: 1,
      moveCount: 0,
    };
  }

  getValidMoves(state) {
    const moves = [];
    for (let from = 0; from < 9; from++) {
      if (state.board[from] !== state.currentPlayer) continue;
      for (const to of this.getDestinations(from, state)) {
        if (state.board[to] === 0) moves.push({ from, to });
      }
    }
    return moves;
  }

  applyMove(state, { from, to }) {
    state.board[to] = state.board[from];
    state.board[from] = 0;
    state.currentPlayer = state.currentPlayer === 1 ? 2 : 1;
    state.moveCount++;
    return state;
  }

  checkResult(state) {
    for (const [a, b, c] of WIN_LINES) {
      const v = state.board[a];
      if (v !== 0 && v === state.board[b] && v === state.board[c]) {
        return { over: true, winner: v, line: [a, b, c] };
      }
    }
    if (state.moveCount >= MAX_MOVES) {
      return { over: true, winner: 0, line: null };
    }
    return { over: false, winner: null, line: null };
  }

  evaluate(state) {
    const result = this.checkResult(state);
    if (!result.over) return 0;
    if (result.winner === 1) return 1000;
    if (result.winner === 2) return -1000;
    return 0;
  }

  // Subclasses implement: return array of position indices this piece can reach
  getDestinations(pos, state) {
    throw new Error('getDestinations() must be implemented by subclass');
  }
}

File: frontend/api/games/movement_game_base.py
"""
Base class for movement-only games (Games 2-5).
Pieces are pre-positioned. No placement phase.

Converted from js/games/MovementGameBase.js
"""
from engine.game_base import GameBase

WIN_LINES = [
    [0, 1, 2], [3, 4, 5], [6, 7, 8],
    [0, 3, 6], [1, 4, 7], [2, 5, 8],
    [0, 4, 8], [2, 4, 6],
]

# P1 at b3(1), a1(6), c1(8); P2 at a3(0), c3(2), b1(7); empty: a2(3), b2(4), c2(5)
INITIAL_BOARD = [2, 1, 2, 0, 0, 0, 1, 2, 1]

MAX_MOVES = 120


class MovementGameBase(GameBase):
    """Base for games where pieces are pre-positioned and only move."""

    @property
    def interaction_mode(self) -> str:
        return "movement"

    @property
    def board_config(self) -> dict:
        return {"type": "tapatan", "size": 3}

    def get_initial_state(self) -> dict:
        return {
            "board": list(INITIAL_BOARD),
            "current_player": 1,
            "move_count": 0,
        }

    def get_valid_moves(self, state: dict) -> list:
        moves = []
        for from_pos in range(9):
            if state["board"][from_pos] != state["current_player"]:
                continue
            for to_pos in self.get_destinations(from_pos, state):
                if state["board"][to_pos] == 0:
                    moves.append({"from": from_pos, "to": to_pos})
        return moves

    def apply_move(self, state: dict, move: dict) -> dict:
        from_pos = move["from"]
        to_pos = move["to"]
        state["board"][to_pos] = state["board"][from_pos]
        state["board"][from_pos] = 0
        state["current_player"] = 2 if state["current_player"] == 1 else 1
        state["move_count"] += 1
        return state

    def check_result(self, state: dict) -> dict:
        board = state["board"]
        for a, b, c in WIN_LINES:
            v = board[a]
            if v != 0 and v == board[b] == board[c]:
                return {"over": True, "winner": v, "line": [a, b, c]}
        if state["move_count"] >= MAX_MOVES:
            return {"over": True, "winner": 0, "line": None}
        return {"over": False, "winner": None, "line": None}

    def evaluate(self, state: dict) -> int:
        result = self.check_result(state)
        if not result["over"]:
            return 0
        if result["winner"] == 1:
            return 1000
        if result["winner"] == 2:
            return -1000
        return 0

    def get_destinations(self, pos: int, state: dict) -> list[int]:
        """Subclasses must implement: return list of reachable position indices."""
        raise NotImplementedError("get_destinations() must be implemented by subclass")
